File: neurokairos/clock_table.py
import datetime as _dt
import json as _json
import warnings
from dataclasses import dataclass, field
from pathlib import Path
from typing import Optional, Union

import numpy as np

_EXTRAP_LIMIT_S = 1.5


@dataclass
class ClockTable:
    """Sparse mapping between source domain (e.g. sample indices) and reference
    domain (e.g. UTC unix timestamps).

    Both arrays must be float64, monotonically increasing, and the same length
    (minimum 2 entries).  Interpolation is done via ``np.interp``.
    """

    source: np.ndarray
    reference: np.ndarray
    nominal_rate: float
    source_units: Optional[str] = None
    metadata: Optional[dict] = field(default=None, repr=False)
    sync_stratum: Optional[np.ndarray] = field(default=None, repr=False)
    sync_dispersion_upperbound_ms: Optional[np.ndarray] = field(
        default=None, repr=False
    )

    def __post_init__(self):
        self.source = np.asarray(self.source, dtype=np.float64)
        self.reference = np.asarray(self.reference, dtype=np.float64)
        if len(self.source) < 2:
            raise ValueError("ClockTable requires at least 2 entries")
        if len(self.source) != len(self.reference):
            raise ValueError("source and reference must have the same length")
        if not np.all(np.diff(self.source) > 0):
            raise ValueError("source must be monotonically increasing")
        if not np.all(np.diff(self.reference) > 0):
            raise ValueError("reference must be monotonically increasing")
        if self.metadata is not None and not isinstance(self.metadata, dict):
            raise TypeError(
                f"metadata must be a dict or None, got {type(self.metadata).__name__}"
            )

        # Validate sync arrays: must be provided as a pair and match source length
        has_stratum = self.sync_stratum is not None
        has_disp = self.sync_dispersion_upperbound_ms is not None
        if has_stratum != has_disp:
            raise ValueError(
                "Must provide both sync_stratum and "
                "sync_dispersion_upperbound_ms, or neither"
            )
        if has_stratum:
            self.sync_stratum = np.asarray(self.sync_stratum, dtype=np.float64)
            self.sync_dispersion_upperbound_ms = np.asarray(
                self.sync_dispersion_upperbound_ms, dtype=np.float64
            )
            if len(self.sync_stratum) != len(self.source):
                raise ValueError(
                    f"sync_stratum length ({len(self.sync_stratum)}) must "
                    f"match source length ({len(self.source)})"
                )
            if len(self.sync_dispersion_upperbound_ms) != len(self.source):
                raise ValueError(
                    f"sync_dispersion_upperbound_ms length "
                    f"({len(self.sync_dispersion_upperbound_ms)}) must "
                    f"match source length ({len(self.source)})"
                )

    def source_to_reference(self, values) -> np.ndarray:
        """Convert source-domain values to reference-domain via interpolation.

        Linearly extrapolates up to 1.5 s beyond the ClockTable boundaries
        (one IRIG-H pulse interval).  Beyond that, returns NaN and issues
        a warning.
        """
        values_arr = np.asarray(values, dtype=np.float64)
        scalar = values_arr.ndim == 0
        v = np.atleast_1d(values_arr)
        result = np.interp(v, self.source, self.reference)

        below = v < self.source[0]
        if np.any(below):
            slope = (self.reference[1] - self.reference[0]) / (
                self.source[1] - self.source[0]
            )
            extrap = self.reference[0] + (v[below] - self.source[0]) * slope
            dist = self.reference[0] - extrap
            # Apply extrapolated values for within-limit entries
            within = dist <= _EXTRAP_LIMIT_S
            result[below] = np.where(within, extrap, np.nan)
            if not np.all(within):
                max_dist = float(np.max(dist))
                warnings.warn(
                    f"source_to_reference: extrapolating {max_dist:.1f} s before "
                    f"the first ClockTable entry exceeds the {_EXTRAP_LIMIT_S} s "
                    f"limit. Returning NaN.",
                    stacklevel=2,
                )

        above = v > self.source[-1]
        if np.any(above):
            slope = (self.reference[-1] - self.reference[-2]) / (
                self.source[-1] - self.source[-2]
            )
            extrap = self.reference[-1] + (v[above] - self.source[-1]) * slope
            dist = extrap - self.reference[-1]
            # Apply extrapolated values for within-limit entries
            within = dist <= _EXTRAP_LIMIT_S
            result[above] = np.where(within, extrap, np.nan)
            if not np.all(within):
                max_dist = float(np.max(dist))
                warnings.warn(
                    f"source_to_reference: extrapolating {max_dist:.1f} s beyond "
                    f"the last ClockTable entry exceeds the {_EXTRAP_LIMIT_S} s "
                    f"limit. Returning NaN.",
                    stacklevel=2,
                )

        return result[0] if scalar else result

    def reference_to_source(self, values) -> np.ndarray:
        """Convert reference-domain values to source-domain via interpolation.

        Linearly extrapolates up to 1.5 s beyond the ClockTable boundaries
        (one IRIG-H pulse interval).  Beyond that, returns NaN and issues
        a warning.
        """
        values_arr = np.asarray(values, dtype=np.float64)
        scalar = values_arr.ndim == 0
        v = np.atleast_1d(values_arr)
        result = np.interp(v, self.reference, self.source)

        below = v < self.reference[0]
        if np.any(below):
            slope = (self.source[1] - self.source[0]) / (
                self.reference[1] - self.reference[0]
            )
            extrap = self.source[0] + (v[below] - self.reference[0]) * slope
            dist = self.reference[0] - v[below]
            # Apply extrapolated values for within-limit entries
            within = dist <= _EXTRAP_LIMIT_S
            result[below] = np.where(within, extrap, np.nan)
            if not np.all(within):
                max_dist = float(np.max(dist))
                warnings.warn(
                    f"reference_to_source: extrapolating {max_dist:.1f} s before "
                    f"the first ClockTable entry exceeds the {_EXTRAP_LIMIT_S} s "
                    f"limit. Returning NaN.",
                    stacklevel=2,
                )

        above = v > self.reference[-1]
        if np.any(above):
            slope = (self.source[-1] - self.source[-2]) / (
                self.reference[-1] - self.reference[-2]
            )
            extrap = self.source[-1] + (v[above] - self.reference[-1]) * slope
            dist = v[above] - self.reference[-1]
            # Apply extrapolated values for within-limit entries
            within = dist <= _EXTRAP_LIMIT_S
            result[above] = np.where(within, extrap, np.nan)
            if not np.all(within):
                max_dist = float(np.max(dist))
                warnings.warn(
                    f"reference_to_source: extrapolating {max_dist:.1f} s beyond "
                    f"the last ClockTable entry exceeds the {_EXTRAP_LIMIT_S} s "
                    f"limit. Returning NaN.",
                    stacklevel=2,
                )

        return result[0] if scalar else result

    def save(self, path: Union[str, Path]) -> None:
        """Save to an NPZ file."""
        path = Path(path)
        arrays = dict(
            source=self.source,
            reference=self.reference,
            nominal_rate=np.array(self.nominal_rate),
        )
        if self.source_units is not None:
            arrays["source_units"] = np.array(self.source_units)
        if self.metadata is not None:
            try:
                arrays["_metadata"] = np.array(_json.dumps(self.metadata))
            except TypeError as e:
                raise TypeError(
                    f"metadata values must be JSON-serializable: {e}"
                ) from e
        if self.sync_stratum is not None:
            arrays["sync_stratum"] = self.sync_stratum
            arrays["sync_dispersion_upperbound_ms"] = (
                self.sync_dispersion_upperbound_ms
            )
        np.savez(path, **arrays)

    @classmethod
    def load(cls, path: Union[str, Path]) -> "ClockTable":
        """Load from an NPZ file."""
        path = Path(path)
        data = np.load(path)
        source_units = str(data["source_units"]) if "source_units" in data else None
        metadata = None
        if "_metadata" in data:
            metadata = _json.loads(str(data["_metadata"]))
        sync_stratum = (
            data["sync_stratum"] if "sync_stratum" in data else None
        )
        sync_disp = (
            data["sync_dispersion_upperbound_ms"]
            if "sync_dispersion_upperbound_ms" in data
            else None
        )
        return cls(
            source=data["source"],
            reference=data["reference"],
            nominal_rate=float(data["nominal_rate"]),
            source_units=source_units,
            metadata=metadata,
            sync_stratum=sync_stratum,
            sync_dispersion_upperbound_ms=sync_disp,
        )

    def __repr__(self):
        units = f" ({self.source_units})" if self.source_units else ""
        lines = [
            f"ClockTable: {len(self.source)} entries{units}, "
            f"rate={self.nominal_rate:.1f}",
        ]

        # Recording start/stop from metadata, falling back to reference array
        meta = self.metadata or {}
        start_str = meta.get("recording_start")
        stop_str = meta.get("recording_stop")

        if start_str is None:
            try:
                dt = _dt.datetime.fromtimestamp(
                    self.reference[0], tz=_dt.timezone.utc
                )
                start_str = dt.strftime("%Y-%m-%dT%H:%M:%SZ")
            except (OSError, OverflowError, ValueError):
                pass
        if stop_str is None:
            try:
                dt = _dt.datetime.fromtimestamp(
                    self.reference[-1], tz=_dt.timezone.utc
                )
                stop_str = dt.strftime("%Y-%m-%dT%H:%M:%SZ")
            except (OSError, OverflowError, ValueError):
                pass

        if start_str and stop_str:
            lines.append(f"  recording: {start_str} → {stop_str}")
        elif start_str:
            lines.append(f"  recording start: {start_str}")

        # File info from metadata
        source_file = meta.get("source_file")
        source_path = meta.get("source_path")
        if source_file and source_path:
            lines.append(f"  file: {source_file} ({source_path})")
        elif source_file:
            lines.append(f"  file: {source_file}")

        # Sync status summary when per-pulse arrays are present
        if self.sync_stratum is not None:
            valid = ~np.isnan(self.sync_stratum)
            if np.any(valid):
                s_min = int(np.nanmin(self.sync_stratum))
                s_max = int(np.nanmax(self.sync_stratum))
                d_min = np.nanmin(self.sync_dispersion_upperbound_ms)
                d_max = np.nanmax(self.sync_dispersion_upperbound_ms)
                s_str = str(s_min) if s_min == s_max else f"{s_min}\u2013{s_max}"
                d_str = (
                    f"< {d_min} ms"
                    if d_min == d_max
                    else f"< {d_min}\u2013{d_max} ms"
                )
                lines.append(f"  sync: stratum {s_str}, dispersion {d_str}")

        lines.append(
            f"  source=[{self.source[0]:.1f}..{self.source[-1]:.1f}], "
            f"reference=[{self.reference[0]:.1f}..{self.reference[-1]:.1f}]"
        )

        return "\n".join(lines)

File: matlab/ClockTable.m
classdef ClockTable
%CLOCKTABLE  Sparse mapping between source and reference clock domains.
%
%   Loads NeuroKairos .clocktable.npz files and provides bidirectional
%   interpolation matching the Python ClockTable behavior exactly.
%
%   ct = ClockTable.load('recording.clocktable.npz');
%   utc = ct.source_to_reference(sample_indices);
%   samples = ct.reference_to_source(utc_timestamps);

    properties (SetAccess = immutable)
        source              % float64 column vector
        reference           % float64 column vector
        nominal_rate        % scalar double
        source_units        % string or ''
        metadata            % struct or []
        sync_stratum        % float64 column vector or []
        sync_dispersion_upperbound_ms  % float64 column vector or []
    end

    properties (Constant, Access = private)
        EXTRAP_LIMIT_S = 1.5;  % max extrapolation distance (seconds)
    end

    methods
        function obj = ClockTable(s)
        %CLOCKTABLE  Construct from struct (as returned by read_npz).
            obj.source = s.source(:);
            obj.reference = s.reference(:);
            obj.nominal_rate = s.nominal_rate;
            obj.source_units = s.source_units;
            obj.metadata = s.metadata;

            if isfield(s, 'sync_stratum')
                obj.sync_stratum = s.sync_stratum(:);
            else
                obj.sync_stratum = [];
            end
            if isfield(s, 'sync_dispersion_upperbound_ms')
                obj.sync_dispersion_upperbound_ms = ...
                    s.sync_dispersion_upperbound_ms(:);
            else
                obj.sync_dispersion_upperbound_ms = [];
            end
        end

        function result = source_to_reference(obj, values)
        %SOURCE_TO_REFERENCE  Convert source-domain values to reference.
        %
        %   ref = ct.source_to_reference(source_values)
        %
        %   Linearly interpolates within bounds. Extrapolates up to 1.5 s
        %   beyond boundaries. Returns NaN beyond that.
            result = interp1( ...
                obj.source, obj.reference, values(:), 'linear');
            result = obj.extrapolate_below( ...
                result, values(:), obj.source, obj.reference);
            result = obj.extrapolate_above( ...
                result, values(:), obj.source, obj.reference);
            % Preserve input shape
            result = reshape(result, size(values));
        end

        function result = reference_to_source(obj, values)
        %REFERENCE_TO_SOURCE  Convert reference-domain values to source.
        %
        %   src = ct.reference_to_source(reference_values)
        %
        %   Linearly interpolates within bounds. Extrapolates up to 1.5 s
        %   beyond boundaries. Returns NaN beyond that.
            result = interp1( ...
                obj.reference, obj.source, values(:), 'linear');
            result = obj.extrapolate_below( ...
                result, values(:), obj.reference, obj.source);
            result = obj.extrapolate_above( ...
                result, values(:), obj.reference, obj.source);
            % Preserve input shape
            result = reshape(result, size(values));
        end

        function disp(obj)
        %DISP  Formatted display matching Python ClockTable.__repr__.
            n = numel(obj.source);
            if isempty(obj.source_units)
                units_str = '';
            else
                units_str = sprintf(' (%s)', obj.source_units);
            end
            fprintf('ClockTable: %d entries%s, rate=%.1f\n', ...
                n, units_str, obj.nominal_rate);

            fprintf('  source=[%.1f..%.1f], reference=[%.1f..%.1f]\n', ...
                obj.source(1), obj.source(end), ...
                obj.reference(1), obj.reference(end));

            if ~isempty(obj.sync_stratum)
                s_min = min(obj.sync_stratum);
                s_max = max(obj.sync_stratum);
                d_min = min(obj.sync_dispersion_upperbound_ms);
                d_max = max(obj.sync_dispersion_upperbound_ms);
                if s_min == s_max
                    s_str = sprintf('%d', s_min);
                else
                    s_str = sprintf('%d-%d', s_min, s_max);
                end
                if d_min == d_max
                    d_str = sprintf('< %.2g ms', d_min);
                else
                    d_str = sprintf('< %.2g-%.2g ms', d_min, d_max);
                end
                fprintf('  sync: stratum %s, dispersion %s\n', ...
                    s_str, d_str);
            end
        end
    end

    methods (Static)
        function obj = load(path)
        %LOAD  Load a ClockTable from a .clocktable.npz file.
        %
        %   ct = ClockTable.load('recording.clocktable.npz')
            s = nk_internal.read_npz(path);
            obj = ClockTable(s);
        end
    end

    methods (Access = private)
        function result = extrapolate_below(obj, result, v, from, to)
        %EXTRAPOLATE_BELOW  Handle values below the first anchor point.
            mask = isnan(result) & (v < from(1));
            if ~any(mask)
                return;
            end
            slope = (to(2) - to(1)) / (from(2) - from(1));
            extrap = to(1) + (v(mask) - from(1)) * slope;

            % Compute distance in the "to" domain (reference for s2r)
            dist = to(1) - extrap;
            within = dist <= obj.EXTRAP_LIMIT_S;
            extrap(~within) = NaN;
            result(mask) = extrap;
        end

        function result = extrapolate_above(obj, result, v, from, to)
        %EXTRAPOLATE_ABOVE  Handle values above the last anchor point.
            mask = isnan(result) & (v > from(end));
            if ~any(mask)
                return;
            end
            slope = (to(end) - to(end-1)) / (from(end) - from(end-1));
            extrap = to(end) + (v(mask) - from(end)) * slope;

            % Compute distance in the "to" domain
            dist = extrap - to(end);
            within = dist <= obj.EXTRAP_LIMIT_S;
            extrap(~within) = NaN;
            result(mask) = extrap;
        end
    end
end
